File: server/src/main/clj/lof/lab/data/record.clj
(ns lof.lab.data.record
  (:require [clojure.string :as str]))

(defrecord Record [canonical-id authors title year doi url])

(defn normalize
  "Normalize a string: lowercase, remove punctuation, trim"
  [s]
  (if (nil? s)
    ""
    (-> s
        str/lower-case
        (str/replace #"[^a-z0-9]" "")
        str/trim)))

(defn extract-author-lastname
  "Extract last name from author string.
   Handles formats: 'LastName, FirstName' or 'FirstName LastName'"
  [author]
  (if (str/includes? author ",")
    ;; Format: "LastName, FirstName"
    (-> author (str/split #",") first str/trim)
    ;; Format: "FirstName LastName"
    (-> author (str/split #"\s+") last str/trim)))

(defn generate-canonical-id
  "Generate canonical ID from (firstAuthor, year, titleSlug).
   Format: 'lastname-year-first-three-title-words'
   Example: 'voros-2023-cradle-things'"
  [{:keys [authors title year]}]
  (let [;; 1. First author's last name (or 'unknown')
        author-part (if (seq authors)
                      (-> authors first extract-author-lastname normalize)
                      "unknown")

        ;; 2. Year (or 'nodate')
        year-part (if year (str year) "nodate")

        ;; 3. First three words of title (or 'notitle')
        title-part (if (and title (not (str/blank? title)))
                     (->> (str/split title #"\s+")
                          (take 3)
                          (map normalize)
                          (str/join "-"))
                     "notitle")]

    (str/join "-" [author-part year-part title-part])))

(defn make-record
  "Create a Record with auto-generated canonical-id"
  [{:keys [authors title year doi url] :as data}]
  (let [canonical-id (generate-canonical-id data)]
    (->Record canonical-id
              (vec (or authors []))
              title
              year
              doi
              url)))

(comment
  ;; REPL testing
  (def test-record
    (make-record {:authors ["Vörös, Sebastjan"]
                  :title "At the Cradle of Things: The Act of Distinction"
                  :year 2023
                  :doi "10.53765/20512201.30.11.017"
                  :url nil}))

  (:canonical-id test-record)
  ;; => "vrs-2023-at-the-cradle"

  (def test-record2
    (make-record {:authors ["Conrad, Leon"]
                  :title "Laws of Form Online Course"
                  :year 2020
                  :doi nil
                  :url "https://www.youtube.com/..."}))

  (:canonical-id test-record2)
  ;; => "conrad-2020-laws-of-form"
  )

File: server/src/main/java/lof/research/lab/data/Record.java
package lof.research.lab.data;

import com.rpl.rama.RamaSerializable;
import java.util.List;
import java.util.ArrayList;

public class Record implements RamaSerializable {
  public String canonicalId;
  public List<String> authors;
  public String title;
  public Integer year;
  public String doi;
  public String url;

  public Record() {
    // Default constructor required for RamaSerializable
  }

  public Record(List<String> authors, String title, Integer year, String doi, String url) {
    this.authors = authors != null ? new ArrayList<>(authors) : new ArrayList<>();
    this.title = title;
    this.year = year;
    this.doi = doi;
    this.url = url;
    this.canonicalId = generateCanonicalId();
  }

  /**
   * Generate a canonical ID from (firstAuthor, year, titleSlug).
   * Format: "lastname-year-first-three-title-words"
   * Example: "voros-2023-cradle-things"
   */
  public String generateCanonicalId() {
    StringBuilder id = new StringBuilder();

    // 1. First author's last name (or "unknown")
    String authorPart = "unknown";
    if (authors != null && !authors.isEmpty()) {
      String firstAuthor = authors.get(0);
      // Extract last name (assume "LastName, FirstName" or "FirstName LastName" format)
      String[] parts = firstAuthor.split(",");
      if (parts.length > 0) {
        authorPart = parts[0].trim();
      } else {
        String[] spaceParts = firstAuthor.split(" ");
        if (spaceParts.length > 0) {
          authorPart = spaceParts[spaceParts.length - 1];
        }
      }
      authorPart = normalize(authorPart);
    }
    id.append(authorPart);

    // 2. Year (or "nodate")
    id.append("-");
    if (year != null) {
      id.append(year);
    } else {
      id.append("nodate");
    }

    // 3. First three words of title (or "notitle")
    id.append("-");
    if (title != null && !title.isEmpty()) {
      String[] words = title.split("\\s+");
      int wordCount = Math.min(3, words.length);
      for (int i = 0; i < wordCount; i++) {
        if (i > 0) id.append("-");
        id.append(normalize(words[i]));
      }
    } else {
      id.append("notitle");
    }

    return id.toString();
  }

  /**
   * Normalize a string: lowercase, remove punctuation, trim
   */
  private String normalize(String s) {
    if (s == null) return "";
    return s.toLowerCase()
            .replaceAll("[^a-z0-9]", "")
            .trim();
  }

  @Override
  public String toString() {
    return String.format("Record{id='%s', title='%s', authors=%s, year=%d, doi='%s', url='%s'}",
            canonicalId,
            title != null ? title.substring(0, Math.min(50, title.length())) + "..." : "null",
            authors,
            year != null ? year : 0,
            doi != null ? doi : "null",
            url != null ? url : "null");
  }
}
